### PR TITLE
Fix k-means++ sampling of Badge and factorize its gradient embeddings

### LINKED ISSUE #586: BUG: Badge.query can return duplicate indices within a single batch
### Describe the issue:

`Badge.query(..., batch_size=6)` can return the same index twice, so the batch
contains fewer than `batch_size` distinct samples and the AL loop silently gets
fewer labels than it paid for.

Expected 6 distinct indices, got `[2, 9, 8, 9, 6, 7]` — index 9 appears twice.

Versions: `scikit-activeml == 1.0.0` (current `master`), `numpy == 2.4.6`,
`scikit-learn == 1.9.0`.

### Reproduce the bug via a code example:

```python
import numpy as np
from sklearn.linear_model import LogisticRegression
from skactiveml.classifier import SklearnClassifier
from skactiveml.pool import Badge
from skactiveml.utils import MISSING_LABEL

# 12 samples: 0 and 1 are labeled, the 10 candidates come from two repeated positions.
X = np.vstack([np.array([[0.0, 0.0], [1.0, 1.0]]),
               np.tile(np.array([[3.0, 3.0], [7.0, 7.0]]), (5, 1))])
y = np.hstack([[0, 1], np.full(10, MISSING_LABEL)])
clf = SklearnClassifier(LogisticRegression(), classes=[0, 1])

query_indices = Badge(random_state=5).query(X, y, clf, batch_size=6)
print("query_indices =", np.asarray(query_indices).tolist())
print("n_distinct    =", len(set(np.asarray(query_indices).tolist())))
```

### Error message:

```shell
No traceback, the malformed batch is returned silently:


query_indices = [2, 9, 8, 9, 6, 7]
n_distinct    = 5
```

### Context for the issue:

A duplicate means the batch is short: Ask for 6 labels and got 5 samples to
annotate, with no warning. Over an AL loop this quietly changes how much of the
labeling budget each round actually spends, and an annotation pipeline that keys
on the sample index sees the same sample twice.

### LINKED ISSUE #587: BUG: Rows of the utilities returned by Badge.query do not sum to one
### Describe the issue:

With `return_utilities=True`, each row is the k-means++ sampling distribution for
that round, but the `nansum` of some rows is only 0.8 or 0.6 (rows 3 and 5 below).

The remaining entries are wrong too, not just rescaled: the four eligible
candidates in row 3 carry 0.2 instead of 0.25, so the row cannot be repaired by
renormalizing it afterwards.

Versions: `scikit-activeml == 1.0.0` (current `master`), `numpy == 2.4.6`,
`scikit-learn == 1.9.0`.

### Reproduce the bug via a code example:

```python
import numpy as np
from sklearn.linear_model import LogisticRegression
from skactiveml.classifier import SklearnClassifier
from skactiveml.pool import Badge
from skactiveml.utils import MISSING_LABEL

# 12 samples: 0 and 1 are labeled, the 10 candidates come from two repeated positions.
X = np.vstack([np.array([[0.0, 0.0], [1.0, 1.0]]),
               np.tile(np.array([[3.0, 3.0], [7.0, 7.0]]), (5, 1))])
y = np.hstack([[0, 1], np.full(10, MISSING_LABEL)])
clf = SklearnClassifier(LogisticRegression(), classes=[0, 1])

query_indices, utilities = Badge(random_state=0).query(
    X, y, clf, batch_size=6, return_utilities=True
)
print("query_indices =", np.asarray(query_indices).tolist())
for i, row in enumerate(utilities):
    print(f"utilities[{i}] = {np.array2string(row, precision=3)}   "
          f"nansum = {np.nansum(row):.3f}")
```

### Error message:

```shell
No traceback, the malformed rows are returned silently:


query_indices = [2, 7, 4, 11, 6, 5]
utilities[0] = [  nan   nan 0.198 0.002 0.198 0.002 0.198 0.002 0.198 0.002 0.198 0.002]   nansum = 1.000
utilities[1] = [nan nan nan 0.2 0.  0.2 0.  0.2 0.  0.2 0.  0.2]   nansum = 1.000
utilities[2] = [  nan   nan   nan 0.125 0.125 0.125 0.125   nan 0.125 0.125 0.125 0.125]   nansum = 1.000
utilities[3] = [nan nan nan 0.2 nan 0.2 0.  nan 0.  0.2 0.  0.2]   nansum = 0.800
utilities[4] = [  nan   nan   nan 0.167   nan 0.167 0.167   nan 0.167 0.167 0.167   nan]   nansum = 1.000
utilities[5] = [nan nan nan 0.2 nan 0.2 nan nan 0.  0.2 0.  nan]   nansum = 0.600
```

### Context for the issue:

_No response_

### LINKED ISSUE #588: BUG: Badge keeps the origin as a permanent k-means++ center, suppressing candidates near it
### Describe the issue:

BADGE should balance uncertainty and diversity and spread the batch out, but one
whole cluster of candidates — those near the origin of the feature space — is
permanently held down at the ~1e-3 scale, and the batch clumps into the other
clusters.

Three separated candidate clusters below: A (indices 2-5), B (6-9), C (10-13,
near the origin). After A and B have been picked from, C is still completely
uncovered yet sits two orders of magnitude below candidates in A and B that are
right next to an already selected sample.

Versions: `scikit-activeml == 1.0.0` (current `master`), `numpy == 2.4.6`,
`scikit-learn == 1.9.0`.

### Reproduce the bug via a code example:

```python
import numpy as np
np.set_printoptions(precision=4, suppress=True, linewidth=200)

from sklearn.linear_model import LogisticRegression
from skactiveml.classifier import SklearnClassifier
from skactiveml.pool import Badge
from skactiveml.utils import MISSING_LABEL

# 14 samples: 0 and 1 are labeled; clusters A (2-5), B (6-9), C (10-13, near origin).
cluster_a = np.array([[10.0, 10.0], [10.5, 9.5], [9.5, 10.5], [10.2, 10.2]])
cluster_b = np.array([[-10.0, -10.0], [-10.5, -9.5], [-9.5, -10.5], [-10.2, -10.2]])
cluster_c = np.array([[0.2, -0.2], [-0.2, 0.2], [0.1, 0.1], [-0.1, -0.1]])
X = np.vstack([np.array([[3.0, -3.0], [-3.0, 3.0]]), cluster_a, cluster_b, cluster_c])
y = np.hstack([[0, 1], np.full(12, MISSING_LABEL)])
clf = SklearnClassifier(LogisticRegression(), classes=[0, 1])

query_indices, utilities = Badge(random_state=1).query(
    X, y, clf, batch_size=4, return_utilities=True
)
print("query_indices =", np.asarray(query_indices).tolist())
for i, row in enumerate(utilities):
    print(f"utilities[{i}] = {np.array2string(row, precision=4)}")
```

### Error message:

```shell
No traceback. Note the last four columns (cluster C) across all four rounds:


query_indices = [5, 6, 7, 8]
utilities[0] = [   nan    nan 0.152  0.0949 0.0949 0.1581 0.152  0.0949 0.0949 0.1581 0.0001 0.0001 0.     0.    ]
utilities[1] = [   nan    nan 0.0001 0.0161 0.1837    nan 0.2941 0.0161 0.1837 0.306  0.0001 0.0001 0.     0.    ]
utilities[2] = [   nan    nan 0.002  0.2697 0.2262    nan    nan 0.2697 0.2262 0.002  0.0016 0.0016 0.0005 0.0005]
utilities[3] = [   nan    nan 0.0043 0.     0.4911    nan    nan    nan 0.4911 0.0043 0.0036 0.0036 0.0011 0.0011]
```

### Context for the issue:

BADGE is the method I reach for when I want the batch spread over the pool, but
here a whole cluster stays unreachable: cluster C never gets picked and its
candidates stay two orders of magnitude below the rest, even once the other two
clusters are already covered. What decides a candidate's chance seems to be how
close it happens to lie to the **origin** of the feature space, not how far it is
from the samples already selected.

### LINKED ISSUE #589: FR: Badge should not materialize the full gradient embeddings
### Describe the workflow you want to enable

Run `Badge.query` on a pool of the size I actually work with. On
`n = 20,000` candidates with `C = 100` classes and a `d = 768` embedding, one call
needs about 25 GB and 9 s for the selection loop, so the strategy is out of reach
on a normal machine long before the pool gets large.

### Describe your proposed solution

`Badge` builds the full `(n, C·d)` gradient-embedding matrix and runs k-means++
on it, but those embeddings never have to exist. The factorization described in
LabelBench (Zhang et al., 2024, Appendix E) gets the same distances out of the
class-probability part and the feature part directly, and the
[official BADGE reference implementation](https://github.com/JordanAsh/badge/blob/master/query_strategies/badge_sampling.py)
has used it since 2023.

For a batch of size `b` this saves memory and time at once:

* space: `O(n·C·d)` → `O(n·(C+d))`
* time: `O(b·n·C·d)` → `O(b·n·(C+d))`

Measured on the selection loop with `batch_size=10`:

| n | C | d | today | factorized |
|---|---|---|---|---|
| 20,000 | 20 | 256 | 0.62 s / 1.6 GB | 0.0055 s / 6.6 MB |
| 20,000 | 100 | 768 | 9.19 s / 24.6 GB | 0.0151 s / 19 MB |

Per round the two are mathematically the same, so up to floating-point rounding
the strategy keeps returning the same indices and the same utilities for a given
`random_state` — nothing about its behavior changes.

### Describe alternatives you've considered, if relevant

The numbers above are float64 synthetic data, one process, 8 BLAS threads,
`scikit-activeml == 1.0.0`, `numpy == 2.4.6`; time is the wall time of the
selection loop, memory the `tracemalloc` peak inside it. I have a working
implementation and would be glad to open a PR.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by TioSisai**
Closes #586
Closes #587
Closes #588
Closes #589

**Additional context**

`Badge.query` now follows the k-means++ semantics of the BADGE reference
implementation and no longer materializes the gradient embeddings.

**Fixes**
- `D^2` is reset to the distance to the first center, so the origin is no
  longer a permanent ghost center (#588).
- `D^2` of the selected centers is zeroed instead of resetting the running
  minimum with `inf`, so no index is returned twice per batch (#586).
- Each `utilities` row is normalized before the selected samples are
  masked, so its `nansum` is one (#587).
- `batch_size` is clamped with a warning if `candidates` is an index array
  holding labeled samples, which previously crashed on `np.nan`
  probabilities.
- If all gradient embeddings are zero, samples are drawn uniformly at
  random and the first pick is controlled by `random_state` as well.

**Efficiency (#589)**
- `<g_i, g_j> = <q_i, q_j> * <v_i, v_j>` replaces the
  `(n_samples, n_classes * n_features)` matrix, which reduces the
  per-round complexity from `O(n * C * d)` to `O(n * (C + d))`.

**Behavioral changes**
- The selected indices change for a given `random_state`. The
  factorization alone is numerically equivalent, but the fixes above
  necessarily alter the sampling, i.e., the corresponding claim in #589
  holds only for the factorization in isolation.
- `query_indices` is a `numpy.ndarray` instead of a `list`, as documented.
- The embeddings are computed in `float64`, since the accumulation error
  of `float32` is large enough to change the sampling.

**Tests**
The existing cases are kept and one regression test per issue is added.
`test_query_matches_reference_kmeanspp` compares every round's utilities
against explicitly materialized embeddings with `rtol=1e-9`.

### CODE REVIEW

**Review by mherde — APPROVED** [review]
Thanks for your contribution!  We'll make small modifications regarding the candidates handling since your code showed that we should add a general test to proper handling of already labeled samples in a one-dimensional candidate indices array. In this case, the candidates is authoritative such labeled candidate indices will remain candidates for querying.
